core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -390,16 +397,21 @@ public MemoryBlock allocatePage(long size, MemoryConsumer consumer) {
     try {
       page = memoryManager.tungstenMemoryAllocator().allocate(acquired);
     } catch (OutOfMemoryError e) {
-      logger.warn("Failed to allocate a page ({} bytes), try again.",
-        MDC.of(LogKeys.PAGE_SIZE, acquired));
+      if (!isRetry) {
+        logger.warn("Failed to allocate a page ({} bytes), try again.", e,
+            MDC.of(LogKeys.PAGE_SIZE, acquired));
+      } else {
+        logger.warn("Failed to allocate a page ({} bytes), try again.",
+            MDC.of(LogKeys.PAGE_SIZE, acquired));
+      }
       // there is no enough memory actually, it means the actual free memory is smaller than
       // MemoryManager thought, we should keep the acquired memory.
       synchronized (this) {
         acquiredButNotUsed += acquired;
         allocatedPages.clear(pageNumber);
       }
       // this could trigger spilling to free some pages.
-      return allocatePage(size, consumer);
+      return allocatePage(size, consumer, true);
```

Comment:
not related to this PR, but the retry is infinite?

Reply:
From a retry logic perspective, this part allows for an infinite number of retries.